File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        new_set= set()

        i=0
        max_length =0
        for j in range(len(s)):

            while s[j] in new_set:
                new_set.remove(s[i])
                i+=1
            
            new_set.add(s[j])
            max_length = max(max_length,j-i+1)

        return max_length

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        //use a hashset
        // use sldiing window
        // shrink the window if the character is already in the hashset


        int i=0;
        int max_len=0;
        HashSet<Character> set = new HashSet();
        for(int j=0;j<s.length();j++)
        {   char ch = s.charAt(j);
            while(set.contains(ch))
            {
                set.remove(s.charAt(i)); //shrink the window
                i++;
            }
            set.add(ch);
            max_len=Math.max(j-i+1,max_len);
        }
        return max_len;
        
    }
}